R/mutate_mzroll_list.R

```diff
@@ -50,12 +50,99 @@ floor_peaks <- function(mzroll_list,
       )),
     missing_peaks
   )
-
+  
   mzroll_list$measurements <- completed_peaks
+  
+  return(mzroll_list)
+}
 
+#' Impute missing peaks with provided feature-level imputation values
+#'
+#'@param mzroll_list: data in triple omic structure
+#'@param lod_values: a tibble that maps groupId to log2 feature-level imputation values
+#'@param quant_var: column to use for peak values
+#'@param imputation_sd: standard deviation of Gaussian distribution to use for missing peak imputation
+#'
+#'@return triple omic data with imputed missing peaks
+#'
+#' @examples
+#' mzroll_list_imputed <- impute_missing_peaks(mzroll_list, lod_values, "log2_abundance", 0.15)
+#'
+#'@export
+impute_missing_peaks <- function(mzroll_list,
+                                 lod_values,
+                                 quant_var = "log2_abundance",
+                                 imputation_sd = 0.15) {
+  
+  test_mzroll_list(mzroll_list)
+  
+  valid_quant_var <- setdiff(
+    mzroll_list$design$measurements$variable,
+    c(mzroll_list$design$feature_pk, mzroll_list$design$sample_pk)
+  )
+  
+  checkmate::assertChoice(quant_var, valid_quant_var)
+  checkmate::assertNumeric(mzroll_list$measurements[[quant_var]])
+  
+  ## find missing peaks
+  missing_peaks <- tidyr::expand_grid(
+    groupId = mzroll_list$features$groupId,
+    sampleId = mzroll_list$samples$sampleId) %>%
+    dplyr::anti_join (
+      mzroll_list$measurements,
+      by = c("groupId", "sampleId")
+    )
+  
+  ## impute missing peaks
+  missing_peaks_imputed <- left_join(
+    missing_peaks, 
+    lod_values, 
+    by = c("groupId")) %>%
+    rowwise() %>%
+    mutate(!!rlang::sym(quant_var) := rnorm(1, mean = !!rlang::sym(quant_var)+1, sd = imputation_sd))
+  
+  ## merge measured peaks with imputed peaks
+  completed_peaks <- dplyr::bind_rows(
+    mzroll_list$measurements,
+    missing_peaks_imputed
+  )
+  
+  mzroll_list$measurements <- completed_peaks
   return(mzroll_list)
 }
 
+#' Fill in missing peaks
+#'
+#'@description
+#'If \code{fill_values} is a data frame, this function calls \code{impute_missing_peaks()}.
+#'If it is a numeric vector, 'this function calls \code{floor_peaks()}. Other types are not currently supported
+#'
+#'@param mzroll_list: data in triple omic structure
+#'@param fill_values: either a numeric constant or a tibble that maps groupId to log2 feature-level imputation values
+#'@param quant_var: column to use for peak values
+#'@param imputation_sd: standard deviation of Gaussian distribution to use for missing peak imputation
+#'
+#'@return triple omic data with imputed missing peaks
+#'
+#' @examples
+#' mzroll_list_imputed <- fill_in_missing_peaks(mzroll_list, lod_values, "log2_abundance", 0.15)
```

Comment:
I'm guessing this won't pass R CMD check because the example won't run without defining lod_values. Please make sure that all examples run and the package passes R CMD check with no errors or warnings. You can run this locally or through github actions (I just reenabled that action since it times out if it hasn't run in a while)

Reply:
Corrected examples here: https://github.com/calico/claman/pull/22/commits/abe5470cb0246a9b65962629f7b1012e346f34f4

R CMD still fails due to other issue. I will create another ticket to handle this separately.